hydroflows/workflow/workflow.py

```diff
@@ -177,6 +180,85 @@ def to_snakemake(
+        cwlfile: Path,
+        dryrun: bool = False,
+    ) -> None:
+        """Save the workflow to a CWL workflow.
+
+        Parameters
+        ----------
+        cwlfile : Path
+            Path to the CWL workflow file. CWL files for individual methods will be created in the subfolder <cwlfile>/cwl.
+        dryrun : bool, optional
+            Run the workflow in dryrun mode, by default False
+        """
+        cwlfile = Path(cwlfile).resolve()
+        configfile = cwlfile.with_suffix(".config.yml")
+        # Make sure all necessary folders exist
+        if not (cwlfile.parent / "cwl").exists():
+            (cwlfile.parent / "cwl").mkdir(parents=True)
+
+        template_env = Environment(
+            loader=PackageLoader("hydroflows"), trim_blocks=True, lstrip_blocks=True
+        )
+        template_workflow = template_env.get_template("workflow.cwl.jinja")
+        template_rule = template_env.get_template("rule.cwl.jinja")
+
+        # Make a copy of workflow so any changes CWL parser will make to config, refs do not alter original
+        workflow_copy = deepcopy(self)
+        cwl_workflow = JinjaCWLWorkflow(
+            rules=[JinjaCWLRule(r) for r in workflow_copy.rules], dryrun=dryrun
+        )
+
+        # Write CWL files for the methods
+        for rule in cwl_workflow.rules:
+            _str = template_rule.render(version=__version__, rule=rule)
+            with open(f"{cwlfile.parent}/cwl/{rule.method_name}.cwl", "w") as f:
+                f.write(_str)
+
+        _str = template_workflow.render(
+            version=__version__,
+            inputs=cwl_workflow.workflow_input,
+            workflow=cwl_workflow,
+            dryrun=dryrun,
+        )
+        with open(cwlfile, "w") as f:
+            f.write(_str)
+
+        # Write CWL config file
+        config = {
+            key: value["value"] for key, value in cwl_workflow.workflow_input.items()
+        }
+        with open(configfile, "w") as f:
+            yaml.dump(config, f)
+
+        # For dryrun, touch missing input files
+        if dryrun:
+            for _, info in config.items():
+                if isinstance(info, dict):
+                    fn = Path(info["path"])
+                    if not fn.is_absolute():
+                        fn = self.root / fn
+                    if not fn.exists():
+                        fn.parent.mkdir(parents=True, exist_ok=True)
+                        fn.touch()
+                elif isinstance(info, list) and all(isinstance(x, dict) for x in info):
+                    for x in info:
+                        fn = Path(x["path"])
+                        if not fn.is_absolute():
+                            fn = self.root / fn
+                        if not fn.exists():
+                            fn.parent.mkdir(parents=True, exist_ok=True)
+                            fn.touch()
+
+        # Copy yml files with nested types
+        yml_root = Path(__file__).parents[1] / "templates"
```

Comment:
It's safer to make a `TEMPLATE_DIR` property in `templates/__init__.py` similar to `cfg/__init__.py` and import it here.